tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala

```diff
@@ -0,0 +1,265 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2016, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.core.tools.linker.backend.emitter
+
+import scala.collection.mutable
+
+import org.scalajs.core.ir.ClassKind
+import org.scalajs.core.ir.Trees.FieldDef
+
+import org.scalajs.core.tools.linker._
+
+private[emitter] object KnowledgeGuardian {
+  private[KnowledgeGuardian] trait Unregisterable {
+    def unregister(invalidatable: Invalidatable): Unit
+  }
+
+  trait Invalidatable {
+    private[KnowledgeGuardian] def registeredTo(
+        unregisterable: Unregisterable): Unit
+
+    def invalidate(): Unit
+  }
+
+  abstract class KnowledgeAccessor(knowledgeGuardian: KnowledgeGuardian)
+      extends GlobalKnowledge with Invalidatable {
+
+    /* In theory, an KnowledgeAccessor should *contain* a GlobalKnowledge, not
+     * *be* a GlobalKnowledge. We organize it that way to reduce memory
+     * footprint and pointer indirections.
+     */
+
+    private val _registeredTo = mutable.Set.empty[Unregisterable]
+
+    def invalidate(): Unit = {
```

Comment:
I'm wondering if we should make this abstract and make a guardian private method that calls `invalidate` and `unregisterFromAll` which we call in this method's stead.
This essentially prevents any mistakes when implementing subclasses of `KnowledgeAccessor`.